src/main/java/com/mojang/serialization/MapCodec.java

```diff
@@ -0,0 +1,289 @@
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.function.UnaryOperator;
+import java.util.stream.Stream;
+
+public abstract class MapCodec<A> extends CompressorHolder implements MapDecoder<A>, MapEncoder<A> {
+    public final <O> RecordCodecBuilder<O, A> forGetter(final Function<O, A> getter) {
+        return RecordCodecBuilder.of(getter, this);
+    }
+
+    public static <A> MapCodec<A> of(final MapEncoder<A> encoder, final MapDecoder<A> decoder) {
+        return of(encoder, decoder, "MapCodec[" + encoder + " " + decoder + "]");
+    }
+
+    public static <A> MapCodec<A> of(final MapEncoder<A> encoder, final MapDecoder<A> decoder, final String name) {
+        return new MapCodec<A>() {
+            @Override
+            public <T> Stream<T> keys(final DynamicOps<T> ops) {
+                return Stream.concat(encoder.keys(ops), decoder.keys(ops));
+            }
+
+            @Override
+            public <T> DataResult<A> decode(final DynamicOps<T> ops, final MapLike<T> input) {
+                return decoder.decode(ops, input);
+            }
+
+            @Override
+            public <T> RecordBuilder<T> encode(final A input, final DynamicOps<T> ops, final RecordBuilder<T> prefix) {
+                return encoder.encode(input, ops, prefix);
+            }
+
+            @Override
+            public String toString() {
+                return name;
+            }
+        };
+    }
+
+    public MapCodec<A> fieldOf(final String name) {
+        return codec().fieldOf(name);
+    }
+
+    @Override
+    public MapCodec<A> withLifecycle(final Lifecycle lifecycle) {
+
+        return new MapCodec<A>() {
+            @Override
+            public <T> Stream<T> keys(final DynamicOps<T> ops) {
+                return MapCodec.this.keys(ops);
+            }
+
+            @Override
+            public <T> DataResult<A> decode(final DynamicOps<T> ops, final MapLike<T> input) {
+                return MapCodec.this.decode(ops, input).setLifecycle(lifecycle);
+            }
+
+            @Override
+            public <T> RecordBuilder<T> encode(final A input, final DynamicOps<T> ops, final RecordBuilder<T> prefix) {
+                return MapCodec.this.encode(input, ops, prefix).setLifecycle(lifecycle);
+            }
+
+            @Override
+            public String toString() {
+                return MapCodec.this.toString();
+            }
+        };
+    }
+
+    public static final class MapCodecCodec<A> implements Codec<A> {
+        private final MapCodec<A> codec;
+
+        public MapCodecCodec(final MapCodec<A> codec) {
+            this.codec = codec;
+        }
+
+        public MapCodec<A> codec() {
+            return codec;
+        }
+
+        @Override
+        public <T> DataResult<Pair<A, T>> decode(final DynamicOps<T> ops, final T input) {
+            return codec.compressedDecode(ops, input).map(r -> Pair.of(r, input));
```

Comment:
As I understand, second argument in `Pair` is supposed to be unparsed element, but right here we get full input.

Reply:
Semantics for MapCodec are "peek at the next map" instead of "consume next map" for legacy reasons, will probably change in the future but not now